collections/permutations.ts

```diff
@@ -20,45 +20,43 @@
  * ])
  * ```
  */
-export function permutations<T>(array: readonly T[]): T[][] {
+export function permutations<T>(inputArray: readonly T[]): T[][] {
   const ret: T[][] = [];
+  const k = inputArray.length;
 
-  if (array.length === 0) {
+  if (k === 0) {
     return ret;
   }
 
-  // Heap Algorithm
-  function heapPermutations(k: number, array: T[]) {
-    const c = new Array<number>(k).fill(0);
+  // Heap's Algorithm
+  const array = [...inputArray];
```

Comment:
This should probably be

```suggestion
  const array = Array.from(inputArray);
```

instead - spreading is not free, so I try to avoid spreading arbitrarily large arrays.

Reply:
Do you have a source for your claim?
1. AFAIK it is the reverse, spreading is almost always faster than `Array.from`. (and in general, things built into the syntax will always be faster than a prototype lookup if they do the same thing)
2. This is an O(n!) operation, you should consider another language/tool if you have more than a dozen elements. (it actually core dump here with 20 elements :sweat_smile:)
3. I am fine changing it to `Array.prototype.slice` through if we are into this kind of optimisation however `slice` preserves holes (:confused:) and works only on array.

Reply:
That was a fun read indeed :D

The post suggests that the spread optimizations apply to Array.from as well. So I gueds it does bot really matter and I do not care either way^^